File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/DmgFileMetaInfo.kt
package com.jetbrains.signatureverifier.serialization

import com.jetbrains.signatureverifier.dmg.DmgFile
import com.jetbrains.signatureverifier.macho.CSMAGIC
import com.jetbrains.signatureverifier.macho.MachoConsts
import com.jetbrains.util.*
import kotlinx.serialization.Serializable
import java.nio.ByteBuffer
import java.nio.channels.SeekableByteChannel

@Serializable
data class DmgFileMetaInfo(
  val fileSize: Long,
  val codeSignaturePointer: DmgFile.CodeSignaturePointer,
  val codeSignatureInfo: CodeSignatureInfo = CodeSignatureInfo()
) : FileMetaInfo {
  override fun modifyFile(stream: SeekableByteChannel, signature: ByteArray) {
    val reader = BinaryReader(stream.Rewind())
    stream.Jump(stream.size() - DmgFile.UDIFResourceFileSize)
    val unsignedUDIFResourceFile = reader.ReadBytes(DmgFile.UDIFResourceFileSize)

    stream.Seek(0, SeekOrigin.End)
    if (fileSize > stream.size()) {
      stream.write(ByteBuffer.wrap(ByteArray((fileSize - stream.size()).toInt() + 1)))
    }

    stream.Jump(codeSignaturePointer.offset)
    stream.write(ByteBuffer.wrap(codeSignatureInfo.toByteArray()))
    codeSignatureInfo.blobs.forEach {
      stream.Jump(codeSignatureInfo.superBlobStart)
      stream.Seek(it.offset.toLong(), SeekOrigin.Current)

      it.length += 2 * UInt.SIZE_BYTES // in dmg `length` includes magicValue and length

      if (it.magic == CSMAGIC.CMS_SIGNATURE) {
        it.content = signature
      }

      if (it.type.toLong() == MachoConsts.CSSLOT_CODEDIRECTORY) {
        stream.write(ByteBuffer.wrap(it.content))
      } else {
        stream.write(ByteBuffer.wrap(it.toByteArray()))
      }
    }

    stream.Jump(stream.size() - DmgFile.UDIFResourceFileSize)
    stream.write(ByteBuffer.wrap(unsignedUDIFResourceFile))

    stream.Jump(stream.size() - DmgFile.UDIFResourceFileSize + DmgFile.codeSignaturePointerOffset)
    stream.write(ByteBuffer.wrap(codeSignaturePointer.toByteArray()))
  }
}

File: net/JetBrains.Serialization/src/FileInfos/Dmg/DmgFileMetaInfo.cs
using JetBrains.FormatRipper.Dmg;
using Newtonsoft.Json;

namespace JetBrains.Serialization.FileInfos.Dmg;

[JsonObject(MemberSerialization.OptIn)]
public class DmgFileMetaInfo : IFileMetaInfo
{
  [JsonProperty("Metadata")] private DmgFileMetadata _metadata;

  [JsonConstructor]
  public DmgFileMetaInfo(DmgFileMetadata metadata)
  {
    _metadata = metadata;
  }

  public unsafe void ModifyFile(Stream stream, byte[] signature)
  {
    stream.Position = stream.Length - sizeof(UDIFResourceFile);
    byte[] unsignedUdifResourceFileBytes = new byte[sizeof(UDIFResourceFile)];
    stream.Read(unsignedUdifResourceFileBytes, 0, sizeof(UDIFResourceFile));

    if (_metadata.FileSize > stream.Length)
    {
      stream.Position = stream.Length;
      stream.Write(new byte[(int)(_metadata.FileSize - stream.Length)], 0,
        (int)(_metadata.FileSize - stream.Length));
    }

    stream.Position = _metadata.CodeSignaturePointer.Position;
    var codeSignatureInfoBytes = _metadata.CodeSignatureInfo.ToByteArray();
    stream.Write(codeSignatureInfoBytes, 0, codeSignatureInfoBytes.Length);

    foreach (var blob in _metadata.CodeSignatureInfo.Blobs)
    {
      stream.Position = _metadata.CodeSignatureInfo.SuperBlobStart + blob.Offset;

      blob.Length += 2 * sizeof(UInt32);

      if (blob.Magic == CSMAGIC_CONSTS.CMS_SIGNATURE)
      {
        blob.Content = signature;
      }

      if (blob.Type == (uint)CSMAGIC_CONSTS.CODEDIRECTORY)
      {
        stream.Write(blob.Content, 0, blob.Content.Length);
      }
      else
      {
        var blobBytes = blob.ToByteArray();
        stream.Write(blobBytes, 0, blobBytes.Length);
      }
    }

    stream.Position = stream.Length - sizeof(UDIFResourceFile);
    stream.Write(unsignedUdifResourceFileBytes, 0, unsignedUdifResourceFileBytes.Length);

    var codeSignaturePointerBytes = _metadata.CodeSignaturePointer.ToByteArray(isBe: true);
    stream.Position = stream.Length - sizeof(UDIFResourceFile) + DmgFile.CODE_SIGNATURE_POINTER_OFFSET;
    stream.Write(codeSignaturePointerBytes, 0, codeSignaturePointerBytes.Length);
  }
}